### ISSUE TITLE
ALiBi with `causal=True` unexpected bias?

### ISSUE THREAD

**Issue by KeremTurgutlu** [mosaicml/llm-foundry#186]
Looking at the [alibi bias generation function](https://github.com/mosaicml/llm-foundry/blob/89f56d2df5f4c0ea10e3209ef0d9bd58dbf05565/llmfoundry/models/layers/attention.py#L572) it seems to me that when `causal=True` a fixed relative position bias (let's ignore head constant for now) is broadcasted to all queries in the q-k attention scores . Let's assume `seq_len=5`:

When we have `full=False`, e.g. `causal=True` and `prefix_lm=False`, we get the following alibi bias before head constant is applied:
 
```python
alibi_bias = torch.arange(1 - seq_len, 1).view(1, 1, 1, seq_len)
alibi_bias, alibi_bias.shape

# shape to be broadcasted is bs x heads x seqlen (query) x seqlen (key)
(tensor([[[[-4, -3, -2, -1,  0]]]]), torch.Size([1, 1, 1, 5]))
```

Reading the AliBi paper my expectation would be an attention bias looking like:

<img width="857" alt="Screenshot 2023-05-21 at 2 22 54 PM" src="https://github.com/mosaicml/llm-foundry/assets/19826777/2bac04d2-cb7d-4075-8bb6-05e397724deb">

```python
# key positions after 0 will be ignored by attn_mask
tensor([[[[ 0, -1, -2, -3, -4],
          [-1,  0, -1, -2, -3],
          [-2, -1,  0, -1, -2],
          [-3, -2, -1,  0, -1],
          [-4, -3, -2, -1,  0]]]])

tensor([[[[ 0, X, X, X, X],
          [-1,  0, X, X, X],
          [-2, -1,  0, X, X],
          [-3, -2, -1,  0, X],
          [-4, -3, -2, -1,  0]]]])
```

When `full=True`, e.g. `causal=False` and `prefix_lm=True` we get the alibi bias below. I think this alibi bias is suitable for any type of model, e.g. CLM(causal lm), PLM(prefix lm), even BERT style mask filling: 

```python
tensor([[[[ 0, -1, -2, -3, -4],
          [-1,  0, -1, -2, -3],
          [-2, -1,  0, -1, -2],
          [-3, -2, -1,  0, -1],
          [-4, -3, -2, -1,  0]]]])
```

1) CLM: Once causal mask is applied positions after 0 will be ignore e.g.:

```python
tensor([[[[ 0, X, X, X, X],
          [-1,  0, X, X, X],
          [-2, -1,  0, X, X],
          [-3, -2, -1,  0, X],
          [-4, -3, -2, -1,  0]]]])
```

2) PLM: Once prefix lm mask is applied with a prefix window 3:

```python
tensor([[[[ 0, -1, -2, X, X],
          [-1,  0, -1, X, X],
          [-2, -1,  0, X, X],
          [-3, -2, -1,  0, X],
          [-4, -3, -2, -1,  0]]]])
```

3) Mask in-filling, no need to apply any mask:

```python
tensor([[[[ 0, -1, -2, -3, -4],
          [-1,  0, -1, -2, -3],
          [-2, -1,  0, -1, -2],
          [-3, -2, -1,  0, -1],
          [-4, -3, -2, -1,  0]]]])
```

**Comment by vchiley** [comment]
`full=True` is the generic correct base case

for CLM, the mask you showed:
```
tensor([[[[ 0,  X,  X,  X,  X],
          [-1,  0,  X,  X,  X],
          [-2, -1,  0,  X,  X],
          [-3, -2, -1,  0,  X],
          [-4, -3, -2, -1,  0]]]])
```
and 
```
tensor([[[[-4,  X,  X,  X,  X],
          [-4, -3,  X,  X,  X],
          [-4, -3, -2,  X,  X],
          [-4, -3, -2, -1,  X],
          [-4, -3, -2, -1,  0]]]])
```
are mathematically equivalent (since softmax(x) = softmax(x + c) for any constant c)
The bias tensor on the bottom is equivalent to having `tensor([[[[-4, -3, -2, -1,  0]]]])` and broadcasting it along the other dimension, then applying causal masking.
Note: this only works if you are applying causal masking.

Attention is pretty BW limited; passing in `tensor([[[[-4, -3, -2, -1,  0]]]])` instead of the full SxS matrix, saves BW and makes the kernel run faster. 
If you are running the attention kernel with alibi, and are using causal masking, then you should use the `tensor([[[[-4, -3, -2, -1,  0]]]])` tensor instead of the full SxS tensor.

This is noted in this thread as well: https://github.com/ofirpress/attention_with_linear_biases/issues/5

**Comment by KeremTurgutlu** [comment]
@vchiley Thanks for the detailed and fast response! Yes, I've missed that they are actually mathematically equivalent. What does BW limited mean? I would only expect a memory change when using a smaller broadcasted `tensor([[[[-4, -3, -2, -1,  0]]]])` but wouldn't imagine speed difference. I will definitely test and time it out, so thanks for pointing it.

For my use case, I plan to experiment with mixture of denoisers similar to UL2 and Palm2 where a batch can have both PLM and CLM. So for that, I set `prefix_lm=True`. For example;

```
# batch size = 4, prefix_mask for the batch
tensor([[1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], # plm with context length = 3
        [1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0.], # plm with context length = 8
        [1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.], # plm with context length = 5
        [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]) # clm
```

I think the symmetric implementation in this repo looks good to me since I plan to use. a decoder only model, thanks for the thread reference.

**Comment by vchiley** [comment]
BW limited == Bandwidth limited
The BW to compute ratio of GPUs means what when a CUDA kernel accesses a float from HBM2 memory, that float should be used about 100 times or else tensorcores will be starved for actual data to process. This is formally known as: BW limited.
If `attn_bias.shape == 1xHx1xS`, when a bias term is accessed it can be reused in **B*S** floating point operations.
if `attn_bias.shape == BxHxSxS`, when a bias term is accessed it can be used in **1** floating point operations.

> but wouldn't imagine speed difference

With `attn_bias.shape == BxHxSxS`, doing some benchmarking with 1B model, I remember going from 50% + MFU to 42% MFU (vs `attn_bias.shape == 1xHx1xS` with broadcasts)
There is a speed difference.

**Comment by KeremTurgutlu** [comment]
I haven't done granular layer-wise profiling but in my case using `prefix mask (bs x L)` vs `causal` is almost same, and even prefix_lm is a bit better. I logged the moving average of throughput for both cases using deepspeed fp16 DDP and MPT-1B model. I also tried FSDP DDP but interestingly it was much slower, but that's another discussion (maybe amp vs fp16 difference - deepspeed could fit bs=16 but fsdp only 12 to a A100-80GB).

Using a simple flag in the same script to test both approaches otherwise everything else is same:

```python
def create_prefix_masks(batch_size, sequence_length, prefix_lengths, device):
    prefix_mask = torch.zeros(batch_size, sequence_length, device=device)
    for pm,n in zip(prefix_mask, prefix_lengths): pm[:n].add_(1)
    return prefix_mask

# Forward pass.
labels = batch.pop("labels")
        
if args.prefix_lm:
    logging.info("Create dummy prefix mask with 0 prefix length - CLM")
    bs = len(batch["input_ids"])
    prefix_mask = create_prefix_masks(bs, args.block_size, prefix_lengths=[0]*bs, device=batch["input_ids"].device)
    outputs = model(**batch, prefix_mask=prefix_mask.bool())
else:
    outputs = model(**batch)
```

```
# prefix lm
steps_per_sec: 1.02 steps_per_sec_per_gpu: 0.51 seqs_per_sec: 32.61 seqs_per_sec_per_gpu: 16.30
# causal
steps_per_sec: 0.93 steps_per_sec_per_gpu: 0.47 seqs_per_sec: 29.90 seqs_per_sec_per_gpu: 14.95
```

**Comment by vchiley** [comment]
The effect I was talking about are only true for `attn_impl: triton`.
`attn_impl: torch` will be slow with either causal or prefixlm.


What is "FSDP DDP"?

**Comment by vchiley** [comment]
btw [this](https://github.com/mosaicml/llm-foundry/pull/193), once merged, will enable torch2 to be used with the triton kernel.

**Comment by KeremTurgutlu** [comment]
> The effect I was talking about are only true for `attn_impl: triton`. `attn_impl: torch` will be slow with either causal or prefixlm.
> 
> What is "FSDP DDP"?

hmm, I see I will try triton impl once it is merged. Is it still based on the hazy implementation? Since PR adds a new file it was difficult to see the diff for changes.

[FSDP](https://huggingface.co/docs/accelerate/usage_guides/fsdp#fully-sharded-data-parallel): Fully Sharded Data Parallel, I used the integration from HF accelerate. DDP is just the distributed data parallel mode of it, meaning that model is copied over each core/gpu. Since it is only 1B params we don't need to shard optimizer, gradients or parameters.

**Comment by KeremTurgutlu** [comment]
Just checked out the latest code and it seems like pinned triton version works with pytorch 2.0. Although, I didn't see much of a speed up just 5% and it is not compatible with `torch.compile` since it is using a custom triton installation now.


```
def timed(fn):
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    result = fn()
    end.record()
    torch.cuda.synchronize()
    return result, start.elapsed_time(end) / 1000

def forward(mod,inp):
    return mod(**inp)

def forward_backward(mod,inp):
    loss = mod(**inp).logits.sum()
    loss.backward()
    return loss

q = torch.randn(4,16,8*128).cuda().half()
k = torch.randn(4,16,8*128).cuda().half()
v = torch.randn(4,16,8*128).cuda().half()

%%timeit
out = triton_flash_attn_fn(q,k,v,n_heads=8)

129 µs ± 138 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%%timeit
out = scaled_multihead_dot_product_attention(q,k,v,n_heads=8)

133 µs ± 102 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

### Test models

# eager model with triton attn
config = MPTConfig()
config.attn_config['attn_impl'] = "triton"
config.attn_config['qk_ln'] = True
config.attn_config['prefix_lm'] = True
model = MPTForCausalLM(config)
model.cuda().half();

bs = 4
seqlen = 256
x = torch.randint(0,config.vocab_size,(bs,seqlen)).cuda()
prefix_mask = torch.zeros(bs, seqlen).cuda().half()
inp = {"input_ids":x, "prefix_mask":prefix_mask}

print("eager/triton[forward]:", timed(lambda: forward(model, inp))[1])
print("eager/triton[forward_backward]:", timed(lambda: forward_backward(model, inp))[1])

eager/triton[forward]: 0.02111692810058594
eager/triton[forward_backward]: 0.062268417358398435

# eager model with torch attn
config = MPTConfig()
config.attn_config['attn_impl'] = 'torch'
config.attn_config['qk_ln'] = True
config.attn_config['prefix_lm'] = True
model = MPTForCausalLM(config)
model.cuda().half();

print("eager/torch[forward]:", timed(lambda: forward(model, inp))[1])
print("eager/torch[forward_backward]:", timed(lambda: forward_backward(model, inp))[1])

eager/torch[forward]: 0.02424831962585449
eager/torch[forward_backward]: 0.06596710205078125
```

Did you benchmarked torch vs triton, how much improvement did you see?

**Comment by vchiley** [comment]
In our experience, `attn_impl: triton` is more memory efficient and is much faster.

**Comment by KeremTurgutlu** [comment]
Thanks for sharing, I would be very interested to see your throughput (e.g. seq / sec / gpu) and wall time numbers for training mpt 1B and 7B models with the hardware specs. Maybe this is a pytorch 2.0 difference I am not sure if they optimized few things and if that is causing this small difference between torch vs triton.

**Comment by vchiley** [comment]
The tables [here](https://github.com/mosaicml/llm-foundry/tree/main/scripts/train/benchmarking#a100-80gb) list the throughput of a large set of networks (using `attn_impl: triton`).
Some of the highlights can be seen [here](https://twitter.com/vitaliychiley/status/1644002362389041153?s=20).

If, using seq len 1024, you can get a 3B model using `attn_impl: torch` to produce almost 62% flop utilization, let me know how you did that 👀 

`attn_impl: torch` uses way more memory and therefore requires using smaller micro batch sizes durring training. Lower micro batch size results in much lower FLOP utilization and throughput.

**Comment by KeremTurgutlu** [comment]
This is very useful thanks for sharing. So far I've only tested with pytorch 2.0 / torch attn impl / deepspeed (zero disabled) DDP fp16 / using 1B model and 1024 seqlen. With this setup I got ~ 14-15 seq/sec/gpu but looking at your table you are able to get ~21 seq/sec/gpu which is pretty impressive. I will run some tests similar to your setup.

**Comment by vchiley** [comment]
Note: those tables use the base model config [here](https://github.com/mosaicml/llm-foundry/blob/main/scripts/train/yamls/pretrain/mpt-1b.yaml) (without alibi or prefix mask)
